gax/src/test/java/com/google/api/gax/batching/v2/FlowControllerTest.java

```diff
@@ -0,0 +1,253 @@
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(10)
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    flowController.reserve(0);
+    Exception actualException = null;
+    try {
+      flowController.reserve(-1);
+      fail("Must have thrown an illegal argument error");
+    } catch (IllegalArgumentException expected) {
+      actualException = expected;
+    }
+    assertThat(actualException).isInstanceOf(IllegalArgumentException.class);
+    flowController.release(1, 0);
+  }
+
+  @Test
+  public void testReserveRelease_noLimits_ok() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    flowController.reserve(1);
+    flowController.release(1, 1);
+  }
+
+  @Test
+  public void testReserveRelease_ignore_ok() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(1)
+                .setMaxOutstandingRequestBytes(1L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    flowController.reserve(1);
+    flowController.release(1, 1);
+  }
+
+  @Test
+  public void testReserveRelease_blockedByElementCount_noBytesLimit() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(2)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+    flowController.reserve(1);
+
+    testBlockingReserveRelease(flowController, 5);
+  }
+
+  @Test
+  public void testReserveRelease_blockedByNumberOfBytes() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(100)
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    testBlockingReserveRelease(flowController, 10);
+  }
+
+  @Test
+  public void testReserveRelease_blockedByNumberOfBytes_noElementCountLimit() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    testBlockingReserveRelease(flowController, 10);
+  }
+
+  private static void testBlockingReserveRelease(
```

Comment:
Generally only tests should be prefixed with `test`, Maybe rename this to assertReserveReleaseBehavior